First Submission: Crossbow

blackDevilNoob opened the pull request:
Overview A fully differentiable, JAX-based implementation of the classic arcade shooter Crossbow. The player controls a cursor to protect a "Friend" character walking across the screen from various environmental threats.

**Core Mechanics**

- Objective: Escort the Friend safely from left to right.
- Combat: Aim and fire to destroy enemies before they touch the Friend.
- Lives: Losing the Friend (who scatters into pixels upon death) costs a life.
- Navigation: Shoot specific targets on the "Map Select" screen to choose the next level.

**Levels & Hazards**

- Desert: Defend against Scorpions, Ants, Snakes, and Vultures.
- Cavern: Survive diving Bats and falling Stalactites.
- Volcano: Dodge gravity-affected Falling Rocks and moving Lava.
- Jungle: Avoid Coconuts dropped by Monkeys and static Voracious Plants.
- Drawbridge: Shoot two ropes to lower the bridge while fighting off Archers.
- Castle Hall: Intercept falling Arrows and tracking Lava projectiles.

**Current Status & Known Bugs**
While the core gameplay loop is functional, there are two known issues in the current build:

1. **Jungle Map Bug:** Killing enemies in the Jungle level may incorrectly kill the Friend character.
2. **Volcano Visuals:** Certain enemies in the Volcano level appear as default placeholders rather than unique rock sprites.

ahmetzahitak committed:
add sprites and update game logic

- add sprites for digits, friend, and first three maps
- update digit, friend, and background rendering to use new assets
- add game phase to keep track of current phase
- update cursor and friend speeds based on the original game
- update the area where cursor can move to match the original game

ahmetzahitak committed:
add sprites

blackDevilNoob committed:
updated desert map

blackDevilNoob committed:
get ready frame with transition updated and player collision scattering added

ahmetzahitak committed:
add three maps and update game phase change logic

blackDevilNoob committed:
demon_map enemies logic added and also refined the enemies logic globally

blackDevilNoob committed:
drawbridge_map enemies logic added and also refined the enemies logic globally and also dessert_map points updated

blackDevilNoob committed:
changed the maped name to castle_Hall_map

Murad2111 committed:
Finalize volcano and jungle map logic

Finalize volcano and jungle map logic

Added new enemy types and corresponding masks:
- Lava rocks
- Falling rocks
- Resting rocks
- Monkey
- Coconut
- Voracious plant

ahmetzahitak committed:
add cave map logic and sprites

ahmetzahitak committed:
Merge branch 'feature/Crossbow' of https://github.com/blackDevilNoob/JAXAtari_Crossbow into feature/Crossbow

ahmetzahitak committed:
Revert to cbd0458: Finalize volcano and jungle map logic

ahmetzahitak committed:
Revert to cbd0458: Finalize volcano and jungle map logic

ahmetzahitak committed:
add cavern and castle hall maps, fix some bugs

blackDevilNoob committed:
solving PR error

blackDevilNoob committed:
converted float to int

blackDevilNoob committed:
added episiode steps

blackDevilNoob committed:
updated env method names

blackDevilNoob committed:
Revert "updated env method names"

This reverts commit a24eb04876e452b9a6f21b6394c3500af60ba8ec.

blackDevilNoob committed:
added reward and done state

Schmitz-David commented:
## Review of First Submission
First of all, I admire your courage to take on such a complex game! Some aspects of the game are replicated quite closely already, and especially the consistent usage of JAX is impressive. You can find my detailed review below.

### Modularity: 4/4
- The code is split very well across many functions which makes it easy to override them when writing game mods.
- From what I've read, the function and variable names are easily understandable and also the structure of the code is reasonable (with well-written comments).

### ALE-similarity: 2.5/4
- The background of the levels are visually indistinguishable from the original game, well done.
- Also, I cannot spot differences in the movement of the crossbow or the 'Friend' characters. The walking animation of the Friends and enemy sprites looks just like in the original.
- The level selection is different: instead of selecting the next level by shooting it, the original uses the coloured buttons below (which also change according to the level, e.g., after the volcano level, there is an additional blue button). However, I actually enjoyed this 'feature' very much during testing, since it makes debugging a lot easier.
- For some levels, the spawning logic is very different, mostly spawning far too many enemies.
- When firing, the cursor becomes red like in the original, but it also becomes slightly larger, which does not happen in the original.
- This 'eye' which sometimes appears and gives 1000 points when shot does not appear if I am not mistaken.
- Detailed breakdown of each level:
	- Volcano level: The volcano does not shoot at all (stationary sprites), and the friend moves across the screen further upwards. Also, the "gap" starts out being hit (visually at least), whereas in the original game the Friend stops once reaching the 'river' if it has not been shot yet. Since the enemies are stationary, no further ones appear if one does not start shooting, making the level very easy to clear by doing nothing. Another point: since the position is chosen randomly, it happens sometimes that the enemy position spawns directly on top of the Friend, so that one loses instantly without having the opportunity to react. Finally, sometimes enemies appear very far down (next to the score), in which case it's not possible to reach them since the crossbow is not allowed to go so far down.
	- Desert level: The enemies look different (e.g., I don't think there are green enemies in the original) and moreover, the scorpions kill the Friend instantly in your game, whereas there is some delay in the original game. In addition, the original game has plants growing (which don't kill the friend but give extra points when shot) which your level is lacking. Finally, in this level, the enemies definitely spawn too quickly.
	- Ice caverns level: The 'gap' in the ground is missing (where you have to shoot the ice crystal on top to continue), but otherwise the level looks very good. I think that the ice sprites continuining all the way to the bottom is not the way it is in the original game. Also, in the original game, the ice crystals appear quite frequently above the Friend, so that they would hit him, and are not completely random. For example, on your level the Friend would be hit be by ~10% of the crystals, in the original game it feels like more than 30%.
	- Drawbridge level: the enemies look different (bats instead of arrows). Also, the arrows move waaay to fast, and the enemies look quite different. Moreover, it's not possible to shoot the bridge ropes, which kind of defeats the purpose of the level. Again, far too many enemies spawn simultaneously (I'm pretty sure the Atari 2600 can't even handle drawing more than 4 simultaneously, and in this level many more appear at once).
	- Castle level: The background and sprites look identical to the original, but I again somehow think that the enemies spawn too quickly.
	- Jungle level: The plants appear instantly without a growing animation. Also, they again spawn too quickly. The projectiles just fall from the top (not being thrown by the monkeys as in the original), and they also fall too quickly if I am not mistaken.

### JAX nativeness: 4/4
- As far as I can see, the code does not make use of any Python control flow and exclusively uses Jax.
- Extremely well-parallelized code which makes use of specialized jnp functions, so that vmap is not even needed (jnp.where, jnp.cumsum, jnp.logical_*, etc.).
- You followed the given framework very closely, which makes the game very compatible to JAXAtari (e.g., _get_default_asset_config).

### Performance: 4/4
- As a consequence of the consistent JAX usage, the performance is great. The game loads quickly, which is impressive in view of its complexity.

ahmetzahitak committed:
add snake and eye sprites

blackDevilNoob committed:
Snake action added

blackDevilNoob committed:
eye action added

Murad2111 committed:
Fix volcano and jungle map logic: bridge mechanic, coconut throwing, collision fixes

blackDevilNoob committed:
Bugs updated

blackDevilNoob committed:
Bugs updated

blackDevilNoob committed:
Globalised it and updated:

1- Changing cursor size and/or speed
2- Changing friend and/or enemy speeds
3- Making the cursor visible only when it is fired
4- Changing the max enemy in an environment
5- Changing enemy attack rate (making the friend die faster or slower)

Murad2111 committed:
Fix lava culling in volcano map: cull lava when it returns to ground level

Murad2111 committed:
Update jungle map: monkeys move vertically, plants rise upward and kill friend

ahmetzahitak committed:
fix cursor movement

- add fire cooldown to prevent continuous firing and moving

ahmetzahitak committed:
add two mods

- cursor only visible when it is fired
- double the max enemies in each map

ahmetzahitak committed:
fix PR test errors

ahmetzahitak committed:
Update core.py

ahmetzahitak committed:
Update jax_crossbow.py

ahmetzahitak committed:
Update jax_crossbow.py

blackDevilNoob committed:
Added 3 mods: FastCursorMod, LargeCursorMod, FastFriendMod

Murad2111 committed:
Add Mod 5: FastEnemyAttackMod and SlowEnemyAttackMod, register crossbow mods in core.py

ahmetzahitak committed:
add flower sprites

ahmetzahitak committed:
Merge branch 'feature/Crossbow' of https://github.com/blackDevilNoob/JAXAtari_Crossbow into feature/Crossbow

Murad2111 committed:
Update jungle map: animated plant sprites, monkeys bounce on ropes, plants static with lifespan

ahmetzahitak committed:
refactor code and organize sprites

ahmetzahitak committed:
fix friend dying logic in desert map

ahmetzahitak committed:
fix volcano map logic

ahmetzahitak committed:
fix flower spawn logic in jungle map

ahmetzahitak committed:
fix rope logic in drawbridge

ahmetzahitak committed:
refactor game logic and map progression

ahmetzahitak committed:
fix player speed

ahmetzahitak committed:
small fixes

ahmetzahitak committed:
update CrossbowObservation

ahmetzahitak committed:
update temporarily

ahmetzahitak committed:
Revert "update temporarily"

This reverts commit 44bf87442ffb66c3af6cfcc0ebb172be5a265d68.

ahmetzahitak committed:
Merge remote-tracking branch 'upstream/master' into feature/Crossbow

ahmetzahitak committed:
revert for PR tests

ahmetzahitak committed:
fix conflicts with dev

ahmetzahitak committed:
fix PR error

ahmetzahitak committed:
Update core.py

ahmetzahitak committed:
minor fix

blackDevilNoob committed:
Increased maximum number of enemies across all maps.
